clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -386,6 +390,13 @@ public long maxOffset() {
             return maxOffset;
         }
 
+        /**
+         * @return  the baseOffset of the first batch of retained records or -1 if no batches are retained
+         */
+        public long baseOffsetOfFirstBatch() {
```

Comment:
In the non-test code, this is only used when we have bytes in the segment (so it will not be -1). In tests, this is sometimes called when the value is -1. I switched to OptionalLong, but we can also switch back.